torch/csrc/jit/script/module.h

```diff
@@ -478,8 +478,8 @@ struct TORCH_API ParameterPolicy {
       IValue v) {
     return std::move(v).toTensor();
   }
-  static bool valid(const ClassTypePtr& typ, size_t i) {
-    return typ->is_parameter(i);
+  static bool valid(const ClassTypePtr& typ, size_t i, const IValue& v) {
+    return typ->is_parameter(i) && v.isTensor();
```

Comment:
I think it's also worth adding some enforcement to `ClassType`: if we're trying to add a parameter, it must be Tensor, Optional[Tensor], or NoneType.